### PR TITLE
add hidden notification count #25

### PR DESCRIPTION
**Pull request by adrubesh**
I went ahead and attempted this. It just simply adds the hidden notification count to the last shown notification in the list. Let me know if there are any issues. 

(I'm Symb0l in the sway-devel irc btw)

### CODE REVIEW

**Comment by emersion** [comment]
Thanks for you PR!

Is it possible to use the same code to draw a notification and to draw the notification count?

File: render.c [hunk 1]
```diff
@@ -114,9 +114,46 @@ int render(struct mako_state *state, struct pool_buffer *buffer) {
 		notif->hotspot.width = notif_width;
 		notif->hotspot.height = notif_height;
 
-		g_object_unref(layout);
-
 		++i;
+		if(i == (size_t)config->max_visible &&
+				i < (size_t)wl_list_length(&state->notifications) ) {
```
Comment (SirCmpwn) [inline render.c]: Style issue, add space before the ( and remove space before the )

**Comment by adrubesh** [comment]
Ahh okay ignore that. Overlooked that option.

**Comment by adrubesh** [comment]
So, I did quite a bit of refactoring. Probably too much for this issue alone but I tried to clean up some of the rendering parts and make some pieces more flexible.

File: include/render.h [hunk 2]
```diff
@@ -3,6 +3,18 @@
 
 struct mako_state;
 
+enum mako_render_text_alignment {
```
Comment (emersion) [inline include/render.h]: No need for this, pango already has it: https://developer.gnome.org/pango/stable/pango-Layout-Objects.html#pango-layout-set-alignment

**Comment by emersion** [comment]
Yeah, I'm not sure I like this. This code will only ever render notifications (ie. boxes with a border and some text inside), so I don't see the value of abstracting things away. Also when we'll add HiDPI support the rendering code will become more complicated to understand, so I really want to keep it as simple as possible.

When I was talking about adding a function that renders notifications, I was thinking of something among the lines of:

```c
int render_notification(struct mako_state *state, int y, const char *text, bool markup);
```

Which would return the height of the renderer box (`notif_height`).

This function can be reused to render the last fake notification at the bottom with the number of hidden notifications.

Don't worry about styling it differently, this will be handled in #3.

File: render.c [hunk 3]
```diff
@@ -97,26 +187,42 @@ int render(struct mako_state *state, struct pool_buffer *buffer) {
...
 		// Update hotspot
 		notif->hotspot.x = 0;
 		notif->hotspot.y = notif_y;
 		notif->hotspot.width = notif_width;
 		notif->hotspot.height = notif_height;
 
-		g_object_unref(layout);
-
 		++i;
+		if (i == (size_t)config->max_visible &&
```
Comment (emersion) [inline render.c]: Can you move this code out of the loop?

**Comment by adrubesh** [comment]
Sorry about that just my coding habbits and I was conflicted with the styling aspect as well. Since that is not an issue, I believe my latest change is more inline with what you wanted.

File: render.c [hunk 4]
```diff
@@ -5,14 +5,83 @@
 #include "notification.h"
 #include "render.h"
 
-static void set_cairo_source_u32(cairo_t *cairo, uint32_t color) {
+
```
Comment (emersion) [inline render.c]: Style nitpick: can you remove these two blank lines?

File: render.c [hunk 5]
```diff
@@ -5,14 +5,83 @@
...
+
+void set_cairo_source_u32(cairo_t *cairo, uint32_t color) {
```
Comment (emersion) [inline render.c]: This function should be `static` because it's only used in this file.

File: render.c [hunk 6]
```diff
@@ -5,14 +5,83 @@
...
 	cairo_set_source_rgba(cairo,
 		(color >> (3*8) & 0xFF) / 255.0,
 		(color >> (2*8) & 0xFF) / 255.0,
 		(color >> (1*8) & 0xFF) / 255.0,
 		(color >> (0*8) & 0xFF) / 255.0);
 }
 
+int render_notification(cairo_t *cairo, struct mako_state *state, const char *text, int offset_y) {
```
Comment (emersion) [inline render.c]: This on should also be `static`

File: render.c [hunk 7]
```diff
@@ -39,89 +105,45 @@ int render(struct mako_state *state, struct pool_buffer *buffer) {
...
-
-		// Render text
-		set_cairo_source_u32(cairo, config->colors.text);
-		cairo_move_to(cairo, config->border_size + config->padding,
-			notif_y + config->border_size + config->padding);
-		pango_cairo_update_layout(cairo, layout);
-		pango_cairo_show_layout(cairo, layout);
+
+		int notif_height = render_notification(cairo, state, text, notif_y);
+		height = notif_y +  notif_height;
```
Comment (emersion) [inline render.c]: Style nitpick: there are two spaces after `+`

File: render.c [hunk 8]
```diff
@@ -39,89 +105,45 @@ int render(struct mako_state *state, struct pool_buffer *buffer) {
...
 			break;
 		}
 	}
 
+	if (wl_list_length(&state->notifications) > config->max_visible) {
+		int hidden = wl_list_length(&state->notifications) - config->max_visible;
+		int hidden_ln = snprintf(NULL, 0, "[%d]", hidden);
+
+		char *hidden_text;
+		hidden_text = malloc(hidden_ln + 1);
```
Comment (emersion) [inline render.c]: This is never free'd. Instead you can use VLAs:

```c
char hidden_text[hidden_ln + 1];
```

File: render.c [hunk 9]
```diff
@@ -39,89 +105,45 @@ int render(struct mako_state *state, struct pool_buffer *buffer) {
...
+		snprintf(hidden_text, hidden_ln + 1, "[%d]", hidden);
+
+		int hidden_height = render_notification(cairo, state, hidden_text, height);
```
Comment (emersion) [inline render.c]: We need to add `inner_margin` to `height` before this line, otherwise the hidden count notification collides with the previous one.

**Comment by adrubesh** [comment]
You have a much better eye for the style issues than I do. Thanks for pointing them out.

File: render.c (hunk 9 above)
Comment (adrubesh) [reply render.c]: The reason for this was that personally I liked the look of the count being closer to the notifications than the spacing provided by `inner_margin` but I added your change.

Comment (emersion) [reply render.c]: Ah, I see. In the future, this can probably be added back as a config option somehow.

**Comment by emersion** [comment]
Opened a follow-up issue: #29

**Pull request merged**